Server/Repositories/TagRepository.cs

```diff
@@ -0,0 +1,171 @@
+﻿using Microsoft.EntityFrameworkCore;
+using Quanda.Server.Data;
+using Quanda.Server.Extensions;
+using Quanda.Shared.DTOs.Requests;
+using Quanda.Shared.DTOs.Responses;
+using Quanda.Shared.Enums;
+using Quanda.Shared.Models;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Quanda.Server.Repositories
+{
+    /// <summary>
+    ///     Repozytorium odpowiedzialne za tagi
+    /// </summary>
+    public interface ITagRepository
+    {
+        /// <summary>
+        ///     Zwraca tagi z podanej strony.
+        /// </summary>
+        /// <param name="page">Aktualna strona przeglądanych tagów.</param>
+        /// <param name="sortOption">Rodzaj sortowania tagów.</param>
+        /// <returns>TagsPageResponseDTO</returns>
+        Task<TagsPageResponseDTO> GetTagsAsync(int page, SortTagsEnum sortOption);
+
+        /// <summary>
+        ///     Zwraca subtagi z podanej strony.
+        /// </summary>
+        /// <param name="idMainTag">Id nadrzędnego tagu.</param>
+        /// <param name="page">Aktualna strona przeglądanych subtagów.</param>
+        /// <param name="sortOption">Rodzaj sortowania tagów.</param>
+        /// <returns>SubTagsPageResponseDTO</returns>
+        Task<SubTagsPageResponseDTO> GetSubTagsAsync(int idMainTag, int page, SortTagsEnum sortOption);
+
+        /// <summary>
+        ///     Zwraca listę tagów danego pytania
+        /// </summary>
+        /// <param name="idQuestion">ID pytania w BD</param>
+        /// <returns>List(TagResponseDTO)</returns>
+        Task<List<TagResponseDTO>> GetTagsOfQuestionAsync(int idQuestion);
+
+        /// <summary>
+        ///     Uaktualnia informacje o tagu
+        /// </summary>
+        /// <param name="tag">Obiekt DTO opisujący tag</param>
+        /// <param name="idTag">Id tagu w BD</param>
+        /// <returns>TagResultEnum</returns>
+        Task<TagResultEnum> UpdateTagAsync(UpdateTagDTO tag, int idTag);
+
+        /// <summary>
+        ///     Dodaję tag do BD
+        /// </summary>
+        /// <param name="category">Obiekt DTO opisujący tag</param>
+        /// <returns>TagResultEnum</returns>
+        Task<TagResultEnum> AddTagAsync(AddTagDTO category);
+
+        /// <summary>
+        ///     Usuwa tag z BD
+        /// </summary>
+        /// <param name="idTag">Id tagu w BD</param>
+        /// <returns>TagResultEnum</returns>
+        Task<TagResultEnum> DeleteTagAsync(int idTag);
+    }
+
+    public class TagRepository : ITagRepository
+    {
+        private readonly AppDbContext _context;
+
+        public TagRepository(AppDbContext context)
+        {
+            _context = context;
+        }
+
+        public async Task<TagsPageResponseDTO> GetTagsAsync(int page, SortTagsEnum sortOption)
+        {
```

Comment:
**This is running synchronously!**
I think this extension method is unnecessary please consider my solution:
![image](https://user-images.githubusercontent.com/53837994/133930595-4007dd7f-82c5-478c-a9c0-2567d7cb1007.png)

Reply:
After this idea with tagsQuery, I agree that extension method is unnecessary. I will removed it.